### PR TITLE
`CommonCircuitData` serialization

### PR DESCRIPTION
**Pull request by Sladuca**
Add support for serializing / deserializing `CommonCircuitData` to the serialization [`Buffer`](https://github.com/mir-protocol/plonky2/blob/dec0765fb554f76e80b8e314b5991318be20609b/plonky2/src/util/serialization.rs#L26).

### CODE REVIEW

File: plonky2/src/gates/gate.rs [hunk 1]
```diff
@@ -17,11 +18,38 @@ use crate::plonk::circuit_builder::CircuitBuilder;
...
+    RandomAccess,
+    ReducingExt,
+    Reducing,
+}
 
 /// A custom gate.
 pub trait Gate<F: RichField + Extendable<D>, const D: usize>: 'static + Send + Sync {
     fn id(&self) -> String;
 
+    fn kind(&self) -> GateKind;
```
Comment (wborgeaud) [inline plonky2/src/gates/gate.rs:46]: One drawback of this approach is that new gates cannot be defined in external crates (as can be seen from the failing test). 
Do you really need this `GateKind` enum in your applications, or would using the `id()` string as a proxy also work?

Comment (wborgeaud) [reply plonky2/src/gates/gate.rs:46]: Maybe we could just have the serialization function take a vector of supported gates `gates: Vec<Box<dyn Gate<F, D>>>` and have the tag of a gate be its index in the vector. That way we don't need the `GateKind` enum nor the `kind()` function. 
Serialization would be dependent on an agreed-upon set of gates for the application, but I think that's a feature as different applications might require different sets of gates.

Comment (Sladuca) [reply plonky2/src/gates/gate.rs:46]: Good idea

Comment (wborgeaud) [reply plonky2/src/gates/gate.rs:46]: Hm good point. It's a bit hacky but I guess we could have `deserialize` be non-static.

**Comment by Sladuca** [comment]
Gonna circle back and do this a different way.

**Pull request closed without merging**